Refactor RobotProperty implementation

xiyuoh opened the pull request:
The current `RobotProperty`/`RobotPropertyKind`/`ModelPropertyData` implementation exist largely within the `widget` module in the site editor. This should not be the case, especially if we want to operate in headless mode. With this PR we move non-UI related logic into their own submodules within the `site` mod.

Also added a fix for the headless roundtrip test failing CI (example https://github.com/open-rmf/rmf_site/pull/345), which occurs because `slotcar` properties are overwriting existing `RobotProperty`/`RobotPropertyKind` imported from file. This is caused by using `commands.entity()` to insert these components during load, but using `world.entity_mut()` when inserting SDF/`slotcar` parameters, so the sequence was messed up. Added some checks in `insert_slotcar_components` to make sure that we're not overwriting existing `RobotProperty` imported from loading.

xiyuoh commented on crates/rmf_site_editor/src/sdf_loader.rs:
```diff
@@ -333,8 +333,12 @@ fn load_model<'a, 'b>(
                         || plugin.filename == "libslotcar.so".to_string()
                     {
+                        // Use commands instead of direct access to world as we
+                        // want this to be queued up after robot data has been
+                        // inserted into model descriptions during loading
                         world
-                            .entity_mut(e)
+                            .commands()
+                            .entity(e)
```
So this is the main change that fixes the "saved robot properties are being overwritten by the SDF loader" bug for me. If I undo this change `cargo test` would fail, and after some logging it turned out the original and destination JSONs are different, with `RobotProperty` being overwritten. This is what I think is happening without `commands()`:

1. `load.rs` [inserts](https://github.com/open-rmf/rmf_site/blob/xiyu/fix_properties/crates/rmf_site_editor/src/site/load.rs#L476-L478) the `Robot` component into relevant model description entities. This would prepare the linked systems to insert appropriate `RobotProperty` ([here](https://github.com/open-rmf/rmf_site/blob/xiyu/fix_properties/crates/rmf_site_editor/src/site/robot_properties.rs#L69)) and `RobotPropertyKind` ([here](https://github.com/open-rmf/rmf_site/blob/xiyu/fix_properties/crates/rmf_site_editor/src/site/robot_properties.rs#L113)) components.
2. Later in the model loading workflow, the SDF loader iterates through all SDF parts (links/visual/etc.) and [inserts](https://github.com/open-rmf/rmf_site/blob/main/crates/rmf_site_editor/src/sdf_loader.rs#L336-L341) the slotcar components into those descendant entities instantly with direct world access. This is because within the SDF loader we don't know what the affiliated descriptions are, since they're not in the same `World` as where everything else is.
3. The [insert_slotcar_components](https://github.com/open-rmf/rmf_site/blob/xiyu/fix_properties/crates/rmf_site_editor/src/site/slotcar.rs#L51) system searches for non-description entities with these `RobotPropertyKind` components, and insert them into the appropriate affiliated description entities. This is done by [writing](https://github.com/open-rmf/rmf_site/blob/xiyu/fix_properties/crates/rmf_site_editor/src/site/slotcar.rs#L160) a `Change` event with the updated `Robot` serialized properties.
4. The ChangePlugin updates `ModelProperty<Robot>` **in the next cycle**, and eventually propagates to the affiliated model instances via the [update_model_instances](https://github.com/open-rmf/rmf_site/blob/xiyu/fix_properties/crates/rmf_site_editor/src/site/model.rs#L873) system.

TLDR I think steps (1) to (3) take place in cycle N, and step (4) takes place in cycle N+1. This means the inserted components in (1) are not known to `insert_slotcar_components` in step (3) and will be overwritten. By using `commands()` we're essentially forcing step 3. to only find slotcar models with robot properties in cycle N+1, and be able to query components from step 1. to check if properties are already present.

xiyuoh replied:
After typing all that out and making changes to `insert_slotcar_components`, doing a local `cargo test` passes with `world.entity_mut()` now. It seems like modifying the query would be a better solution, reverted this change! [19e9e93](https://github.com/open-rmf/rmf_site/pull/385/commits/19e9e93ff53db2226144fbe431fc07d1f8b7138b)

luca-della-vedova replied:
My assertion here is that the world you see here is not the `App`'s world, but a brand new one that is immediately serialized into a scene, and in this process only entities and components are saved, the commands are effectively dropped.
I tested this by actually removing the whole statement that uses a command to add components altogether and verified that the unit test still passes. This makes me think that our unit test doesn't actually verify that this statement is working

xiyuoh commented on crates/rmf_site_editor/src/site/slotcar.rs:
```diff
@@ -94,45 +100,61 @@ fn insert_slotcar_components(
-                        .insert(PowerSource::label(), power_source_value);
-                }
+                // Only allow overwriting RobotProperties that are not present
+                let Ok((mobility, power_source, power_dissipation)) = robot_properties.get(desc)
```
Ah I'm trying to figure out if any combination of `RobotProperty` was already present in the description, and only overwrite components that are not present. So it was intentional to query for optional `RobotProperty` components for all model descriptions.

But on second thought if a particular component is present in a saved JSON and another is not, the latter is likely meant to be disabled. I'll remove the `Option` and check for present `RobotProperty` components as a group.

xiyuoh replied:
Note, `IsStatic` has to be optional else slotcar parameters may not be inserted at all, even if there's no existing params

xiyuoh commented on crates/rmf_site_editor/src/site/robot_properties.rs:
```diff
@@ -28,6 +30,40 @@ pub struct UpdateRobotPropertyKinds {
+        app.init_resource::<ModelPropertyData>()
+            .world_mut()
+            .resource_mut::<ModelPropertyData>()
+            .optional
```
We need the `FromWorld` for this resource and then modify it, so I couldn't build in-place. But I built and inserted it separately anyway [e8f7bf2](https://github.com/open-rmf/rmf_site/pull/385/commits/e8f7bf2a2f354d83d9041fac0ec016e817a5ff67)

xiyuoh commented on crates/rmf_site_editor/src/site/slotcar.rs:
```diff
@@ -94,45 +100,61 @@ fn insert_slotcar_components(
-                {
-                    if let Ok(mechanical_system_value) =
-                        serialize_robot_property_kind::<MechanicalSystem>(mechanical_system.clone())
+                if mobility.is_none() && is_static.get(desc).is_ok_and(|is| !is.0 .0) {
```
[e8f7bf2](https://github.com/open-rmf/rmf_site/pull/385/commits/e8f7bf2a2f354d83d9041fac0ec016e817a5ff67)

xiyuoh replied:
[19e9e93](https://github.com/open-rmf/rmf_site/pull/385/commits/19e9e93ff53db2226144fbe431fc07d1f8b7138b)

xiyuoh commented on crates/rmf_site_editor/src/site/robot_properties.rs:
```diff
@@ -28,6 +30,40 @@ pub struct UpdateRobotPropertyKinds {
+            .world()
+            .components()
+            .component_id::<ModelProperty<Robot>>()
+            .unwrap();
```
[e8f7bf2](https://github.com/open-rmf/rmf_site/pull/385/commits/e8f7bf2a2f354d83d9041fac0ec016e817a5ff67)

xiyuoh commented:
Some changes from the last review (significant enough for a PR title change):
- Use observers instead of scheduled systems to update changes to `ModelProperty<Robot>` and insert/remove relevant components into/from description entities. This should help make sure changes are propagated and present by the next cycle.
- Cleaned up `slotcar` implementation to fix overwriting issue
- Tackled `TODO` for supporting multi-kind robot properties (e.g. `PowerDissipation` that supports multiple robot property kinds, i.e. `AmbientSystem` and `MechanicalSystem`) instead of having to write a separate plugin for these use cases
- Introduced `EmptyRobotProperty<T>` - can be used to represent `RobotProperty` intentionally left empty in the site. Used `EmptyRobotPropertyPlugin::<T>::new()` to enable this for a chosen robot property.
- Moved `RobotProperty` and `RobotPropertyKind` registration logic out of `widget` module and into `site`. This ensures that we're still registering and tracking these properties in headless mode.

luca-della-vedova commented on crates/rmf_site_editor/src/site/robot_properties.rs:
```diff
@@ -28,78 +31,323 @@ pub struct UpdateRobotPropertyKinds {
+            _ignore: Default::default(),
+        }
+    }
+}
```
Nit, is it possible to jut derive `Default` and get rid of this custom implementation since it really just calls `default` for the only member variable in the struct?

luca-della-vedova commented on crates/rmf_site_editor/src/site/robot_properties.rs:
```diff
@@ -28,78 +31,323 @@ pub struct UpdateRobotPropertyKinds {
+                RobotPropertyRegistration {
+                    widget: None,
+                    kinds: HashMap::new(),
+                },
```
I also wonder if it would make sense to derive `Default` for this struct and just insert its default value

luca-della-vedova commented on crates/rmf_site_editor/src/site/robot_properties.rs:
```diff
@@ -28,78 +31,323 @@ pub struct UpdateRobotPropertyKinds {
+            _ignore: Default::default(),
+        }
+    }
+}
```
Same as above as deriving `Default` so we can get rid of this custom implementation

luca-della-vedova commented on crates/rmf_site_editor/src/site/slotcar.rs:
```diff
@@ -56,85 +61,102 @@ fn insert_slotcar_components(
+                    } else {
+                        // We have to insert this ModelProperty as the ChangePlugin does
+                        // not accept changes if the component is not already present
+                        commands.entity(desc).insert(ModelProperty(robot.clone()));
```
I don't know if using the `ChangePlugin` would be better but this comment is not quite correct, the `ChangePlugin` [allows insertion](https://github.com/open-rmf/rmf_site/blob/0ec90fdd3b4344b7c27bb10b3514952fc74d2b1d/crates/rmf_site_editor/src/site/change_plugin.rs#L41-L44) as a (default off) behavior,

luca-della-vedova reviewed (commented):
Understood on the behavior, I left some inline reviews, behavior wise lgtm

xiyuoh commented on crates/rmf_site_editor/src/site/robot_properties.rs:
```diff
@@ -28,78 +31,323 @@ pub struct UpdateRobotPropertyKinds {
+            serde_json::Value::Object(Map::new())
+        }
+        Err(_) => return,
+    };
```
It's possible for there to be other forms of error from `retrieve_robot_property`, e.g. `RobotPropertyError::DeserializePropertyError` if deserializing fails. But you made me realize I should do a printout of the error instead of simply returning. Added error logs in [6e54e79](https://github.com/open-rmf/rmf_site/pull/385/commits/6e54e792cd9a47369e8874a79d9210dca36a23e7).

luca-della-vedova replied:
Ah yes I missed the `map` call :facepalm: thanks for pointing that out

xiyuoh commented on crates/rmf_site_editor/src/site/slotcar.rs:
```diff
@@ -56,85 +61,102 @@ fn insert_slotcar_components(
+        slotcar_property_kinds.iter()
     {
         if !model_descriptions.get(e).is_ok() {
-            // A non-description entity has the RobotPropertyKind component, it could have been inserted into a
```
That's a really good point, this check is redundant since it's satisfied by the QueryFilter. Removed in [6e54e79](https://github.com/open-rmf/rmf_site/pull/385/commits/6e54e792cd9a47369e8874a79d9210dca36a23e7).

luca-della-vedova reviewed (approved):
Thanks for iterating! LGTM
